File: Freestylecoding.Math.FSharp.Tests/Natural/Comparable.fs
namespace Natural

open Xunit
open Freestylecoding.Math

type public Comparable() =
    let eq actual = actual = 0
    let gt actual = actual > 0
    let lt actual = actual < 0

    [<Theory>]
    [<InlineData( 0u, 0u,  0 )>]
    [<InlineData( 1u, 1u,  0 )>]
    [<InlineData( 1u, 0u,  1 )>]
    [<InlineData( 0u, 1u, -1 )>]
    member public this.Sanity left right expected =
        let test =
            match expected with
            | 0 -> eq
            | p when p > 0 -> gt
            | n when n < 0 -> lt
            | _ -> raise (System.Exception( "Not Possible" ))

        Assert.True(
            test ((Natural( [left] ) :> System.IComparable).CompareTo( Natural( [right] )) )
        )

    [<Fact>]
    member public this.Equals () =
        let b = System.Convert.ToByte( System.Random().Next( 0, 256 ) )
        let left = Natural( b ) :> System.IComparable

        Assert.True( eq (left.CompareTo( Natural( b ) )) )
        Assert.True( eq (left.CompareTo( b )) )
        Assert.True( eq (left.CompareTo( uint16 b )) )
        Assert.True( eq (left.CompareTo( uint32 b )) )
        Assert.True( eq (left.CompareTo( uint64 b )) )
        Assert.True( eq (left.CompareTo( System.UInt128( 0uL, uint64 b ) )) )
        Assert.True( eq (left.CompareTo( System.Numerics.BigInteger b )) )

    [<Fact>]
    member public this.GreaterThan () =
        let left = Natural( System.Convert.ToUInt32( System.Random().Next( 128, 256 ) ) ) :> System.IComparable
        let b = System.Convert.ToByte( System.Random().Next( 0, 127 ) )

        Assert.True( gt (left.CompareTo( Natural( b ) )) )
        Assert.True( gt (left.CompareTo( b )) )
        Assert.True( gt (left.CompareTo( uint16 b )) )
        Assert.True( gt (left.CompareTo( uint32 b )) )
        Assert.True( gt (left.CompareTo( uint64 b )) )
        Assert.True( gt (left.CompareTo( System.UInt128( 0uL, uint64 b ) )) )
        Assert.True( gt (left.CompareTo( System.Numerics.BigInteger b )) )

    [<Fact>]
    member public this.GreaterThan_NegativeBigInteger () =
        let left = Natural.Unit :> System.IComparable

        Assert.True( gt (left.CompareTo( System.Numerics.BigInteger( -1m ) )) )

    [<Fact>]
    member public this.LessThan () =
        let left = Natural( System.Convert.ToUInt32( System.Random().Next( 0, 127 ) ) ) :> System.IComparable
        let b = System.Convert.ToByte( System.Random().Next( 128, 256 ) )

        Assert.True( lt (left.CompareTo( Natural( b ) )) )
        Assert.True( lt (left.CompareTo( b )) )
        Assert.True( lt (left.CompareTo( uint16 b )) )
        Assert.True( lt (left.CompareTo( uint32 b )) )
        Assert.True( lt (left.CompareTo( uint64 b )) )
        Assert.True( lt (left.CompareTo( System.UInt128( 0uL, uint64 b ) )) )
        Assert.True( lt (left.CompareTo( System.Numerics.BigInteger b )) )

    [<Fact>]
    member public this.LessThan_NegativeBigInteger () =
        let left = Natural.Unit :> System.IComparable

        Assert.False( lt (left.CompareTo( System.Numerics.BigInteger( -1m ) )) )

    [<Theory>]
    [<InlineData( 1  )>]
    [<InlineData( 1L )>]
    [<InlineData( 1F )>]
    [<InlineData( 1.0 )>]
    [<InlineData( "1" )>]
    member public this.Incompatible value =
        let one = Natural.Unit :> System.IComparable
        let exc = Record.Exception(
            fun () ->
                one.CompareTo( value )
                |> ignore
        )
        Assert.NotNull( exc )
        Assert.IsType<System.ArgumentException>( exc )

File: Freestylecoding.Math.CSharp.Tests/Natural/Comparable.cs
using System;
using Xunit;

namespace Freestylecoding.Math.CSharp.Tests.Natural;

using static Freestylecoding.Math.CSharp.Tests.Helpers;
using Natural = Freestylecoding.Math.Natural;

public class Comparable {
	private static readonly Random random = new Random();

	private static Func<int,bool> eq = (int actual) => actual == 0;
	private static Func<int,bool> gt = (int actual) => actual > 0;
	private static Func<int,bool> lt = (int actual) => actual < 0;

	[Theory]
	[InlineData( 0u, 0u, 0 )]
	[InlineData( 1u, 1u, 0 )]
	[InlineData( 1u, 0u, 1 )]
	[InlineData( 0u, 1u, -1 )]
	public void Sanity( uint left, uint right, int expected ) {
		Func<int, bool> TestFunc = expected switch {
			0 => eq,
			_ when expected > 0 => gt,
			_ when expected < 0 => lt,
			_ => throw new Exception( "Not Possible" )
		};

		Assert.True(
			TestFunc( ( new Natural( left ) as IComparable ).CompareTo( new Natural( right ) ) )
		);
	}

	[Fact( DisplayName = "Equals" )]
	public void EqualsTest() {
		byte b = Convert.ToByte( random.Next( 0, 256 ) );
		IComparable left = new Natural( b ) as IComparable;

		Assert.True( eq( left.CompareTo( new Natural( b ) ) ) );
		Assert.True( eq( left.CompareTo( b ) ) );
		Assert.True( eq( left.CompareTo( (ushort)b ) ) );
		Assert.True( eq( left.CompareTo( (uint)b ) ) );
		Assert.True( eq( left.CompareTo( (ulong)b ) ) );
		Assert.True( eq( left.CompareTo( new UInt128( 0uL, (ulong)b ) ) ) );
		Assert.True( eq( left.CompareTo( new System.Numerics.BigInteger( b ) ) ) );
	}

	[Fact]
	public void GreaterThan() {
		IComparable left = new Natural( Convert.ToUInt32( random.Next( 128, 256 ) ) ) as System.IComparable;
		byte b = Convert.ToByte( random.Next( 0, 127 ) );


		Assert.True( gt( left.CompareTo( new Natural( b ) ) ) );
		Assert.True( gt( left.CompareTo( b ) ) );
		Assert.True( gt( left.CompareTo( (ushort)b ) ) );
		Assert.True( gt( left.CompareTo( (uint)b ) ) );
		Assert.True( gt( left.CompareTo( (ulong)b ) ) );
		Assert.True( gt( left.CompareTo( new UInt128( 0uL, (ulong)b ) ) ) );

		Assert.True( gt( left.CompareTo( new System.Numerics.BigInteger( b ) ) ) );
	}

	[Fact]
	public void GreaterThan_NegativeBigInteger() {
		IComparable left = Natural.Unit as IComparable;
		Assert.True( gt( left.CompareTo( new System.Numerics.BigInteger( -1m ) ) ) );
	}

	[Fact]
	public void LessThan() {
		IComparable left = new Natural( Convert.ToUInt32( random.Next( 0, 127 ) ) ) as IComparable;
		byte b = Convert.ToByte( random.Next( 128, 256 ) );


		Assert.True( lt( left.CompareTo( new Natural( b ) ) ) );
		Assert.True( lt( left.CompareTo( b ) ) );
		Assert.True( lt( left.CompareTo( (ushort)b ) ) );
		Assert.True( lt( left.CompareTo( (uint)b ) ) );
		Assert.True( lt( left.CompareTo( (ulong)b ) ) );
		Assert.True( lt( left.CompareTo( new UInt128( 0uL, (ulong)b ) ) ) );

		Assert.True( lt( left.CompareTo( new System.Numerics.BigInteger( b ) ) ) );
	}

	[Fact]
	public void LessThan_NegativeBigInteger() {
		IComparable left = Natural.Unit as IComparable;
		Assert.False( lt( left.CompareTo( new System.Numerics.BigInteger( -1m ) ) ) );
	}

	[Theory]
	[InlineData( 1 )]
	[InlineData( 1L )]
	[InlineData( 1F )]
	[InlineData( 1.0 )]
	[InlineData( "1" )]
	public void Incompatible( dynamic value ) {
		IComparable one = Natural.Unit as System.IComparable;
		Exception exc = Record.Exception(
			() => one.CompareTo( value )
		);

		Assert.NotNull( exc );
		Assert.IsType<System.ArgumentException>( exc );
	}
}